core/src/main/scala/quasar/destination/s3/DataEventConsumer.scala

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2020 Precog Data
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.destination.s3
+
+import slamdata.Predef._
+
+import quasar.api.push.OffsetKey
+import quasar.connector.DataEvent
+
+import cats.effect.Concurrent
+import cats.syntax.all._
+
+import fs2.{Pipe, Pull, Stream}
+import fs2.concurrent.Queue
+
+class DataEventConsumer[F[_], A, B](queue: Queue[F, Option[OffsetKey.Actual[A]]], write: Pipe[F, B, Unit]) {
+
+  private def go(inp: Stream[F, DataEvent[B, OffsetKey.Actual[A]]]): Pull[F, B, Unit] =
+    inp.pull.uncons1 flatMap {
```

Comment:
Maybe use `flatMap` on `inp`? 

```scala
inp.evalMap {
  case DataEvent.Create(chunk) => Stream.emit(chunk)
  case DataEvent.Delete(_) => Stream.empty
  case DataEvent.Commit(offset) => Stream.eval(queue.enqueue(offset.some)).drain
}
```
It's not any particular better than `Pull` of course :shrug: